File: jcode-panel/jcode_panel/hotkeys.py
from __future__ import annotations

from dataclasses import dataclass
import os
from typing import Callable


MODIFIER_ALIASES = {
    "control": "ctrl",
    "control_l": "ctrl",
    "control_r": "ctrl",
    "ctrl_l": "ctrl",
    "ctrl_r": "ctrl",
    "shift_l": "shift",
    "shift_r": "shift",
    "alt_l": "alt",
    "alt_r": "alt",
    "option": "alt",
    "super_l": "super",
    "super_r": "super",
    "cmd": "super",
    "command": "super",
    "win": "super",
    "windows": "super",
    "esc": "escape",
    "return": "enter",
    "kp_enter": "enter",
}

MODIFIER_ORDER = ["ctrl", "alt", "shift", "super"]
MODIFIERS = set(MODIFIER_ORDER)


@dataclass
class HotkeyStatus:
    enabled: bool
    message: str
    fallback: str = "Use the tray menu → Prompt, or run `jcode-panel --prompt`."


def normalize_key_name(name: str) -> str:
    key = (name or "").strip().lower().replace(" ", "_")
    if key.startswith("key."):
        key = key[4:]
    return MODIFIER_ALIASES.get(key, key)


def key_name_from_pynput(key) -> str:
    name = getattr(key, "name", None) or getattr(key, "char", "") or ""
    return normalize_key_name(str(name))


def normalize_hotkey(hotkey: str) -> str:
    parts = [normalize_key_name(part) for part in str(hotkey or "").replace("-", "+").split("+")]
    parts = [part for part in parts if part]
    modifiers = [mod for mod in MODIFIER_ORDER if mod in parts]
    keys = [part for part in parts if part not in MODIFIERS]
    key = keys[-1] if keys else ""
    return "+".join(modifiers + ([key] if key else [])) or "f8"


def hotkey_parts(hotkey: str) -> tuple[set[str], str]:
    parts = normalize_hotkey(hotkey).split("+")
    modifiers = {part for part in parts if part in MODIFIERS}
    keys = [part for part in parts if part not in MODIFIERS]
    return modifiers, (keys[-1] if keys else "")


def start_hotkey_listener(hotkey: str, callback: Callable[[], None]) -> HotkeyStatus:
    """Start best-effort global hotkey listener.

    On Wayland, global key capture is often blocked. We still try because some
    sessions allow it, but return an explicit degraded-status message.
    """
    session_type = os.environ.get("XDG_SESSION_TYPE", "").lower()
    try:
        from pynput import keyboard  # type: ignore
    except Exception as exc:
        return HotkeyStatus(False, f"Global hotkey unavailable: {exc}")

    normalized_hotkey = normalize_hotkey(hotkey)
    wanted_mods, wanted_key = hotkey_parts(normalized_hotkey)
    pressed_mods: set[str] = set()

    def on_press(key):
        key_name = key_name_from_pynput(key)
        if key_name in MODIFIERS:
            pressed_mods.add(key_name)
        elif key_name == wanted_key and wanted_mods.issubset(pressed_mods):
            callback()

    def on_release(key):
        key_name = key_name_from_pynput(key)
        if key_name in MODIFIERS:
            pressed_mods.discard(key_name)

    try:
        listener = keyboard.Listener(on_press=on_press, on_release=on_release)
        listener.daemon = True
        listener.start()
    except Exception as exc:
        return HotkeyStatus(False, f"Global hotkey failed to start: {exc}")

    if session_type == "wayland":
        return HotkeyStatus(True, "Wayland detected: F8 may be blocked by compositor. Tray Prompt still works.")
    return HotkeyStatus(True, f"Global hotkey active: {normalized_hotkey}")

File: jcode-panel-tauri/src-tauri/src/core/hotkeys.rs
use std::collections::BTreeSet;

const MODIFIER_ORDER: [&str; 4] = ["ctrl", "alt", "shift", "super"];

pub fn normalize_key_name(name: &str) -> String {
    let key = name.trim().to_ascii_lowercase().replace(' ', "_");
    let key = key.strip_prefix("key.").unwrap_or(&key);
    match key {
        "control" | "control_l" | "control_r" | "ctrl_l" | "ctrl_r" => "ctrl".into(),
        "shift_l" | "shift_r" => "shift".into(),
        "alt_l" | "alt_r" | "option" => "alt".into(),
        "super_l" | "super_r" | "cmd" | "command" | "win" | "windows" => "super".into(),
        "esc" => "escape".into(),
        "return" | "kp_enter" => "enter".into(),
        _ => key.into(),
    }
}

pub fn normalize_hotkey(hotkey: &str) -> String {
    let parts: Vec<String> = hotkey
        .replace('-', "+")
        .split('+')
        .map(normalize_key_name)
        .filter(|p| !p.is_empty())
        .collect();
    let modifiers: Vec<&str> = MODIFIER_ORDER
        .iter()
        .copied()
        .filter(|m| parts.iter().any(|p| p == m))
        .collect();
    let key = parts
        .iter()
        .rev()
        .find(|p| !MODIFIER_ORDER.contains(&p.as_str()))
        .cloned()
        .unwrap_or_default();
    let mut out: Vec<String> = modifiers.into_iter().map(str::to_string).collect();
    if !key.is_empty() {
        out.push(key);
    }
    if out.is_empty() {
        "f8".into()
    } else {
        out.join("+")
    }
}

pub fn hotkey_parts(hotkey: &str) -> (BTreeSet<String>, String) {
    let normalized = normalize_hotkey(hotkey);
    let mut mods = BTreeSet::new();
    let mut key = String::new();
    for part in normalized.split('+') {
        if MODIFIER_ORDER.contains(&part) {
            mods.insert(part.to_string());
        } else {
            key = part.to_string();
        }
    }
    (mods, key)
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn normalizes_hotkeys_and_parts() {
        assert_eq!(normalize_key_name("Key.Control_L"), "ctrl");
        assert_eq!(normalize_hotkey("Shift-Ctrl-F8"), "ctrl+shift+f8");
        let (mods, key) = hotkey_parts("ctrl+shift+s");
        assert!(mods.contains("ctrl") && mods.contains("shift"));
        assert_eq!(key, "s");
    }
}
